### DISCUSSION TITLE
Help request (Staubli TX60)

### DISCUSSION THREAD

**Discussion by eduardsik** [visose/Robots#103] [Q&A] [answered]
Hello,

I would like to ask for help with Staubli XT60, to add to the library. I modified the bartlett library and add the robot to xml, and 3d model, but can not fix to not falling apart, even if using the DH parameters. 
May I ask you for add the robot correctly? 
Here are the necessary files (3d model, xml, DH parameters): https://www.dropbox.com/sh/9ziq0qys1qapfl9/AAC4jETHgbto-r9fl0a2YIYma?dl=0
Thank you very mutch!

**Comment by lin-ycv** [comment]
I think the problem is with the plugin itself?
TX60 is slightly special in that not all joints are on the XZ plane, I think this is what's causing problems.
![image](https://user-images.githubusercontent.com/60565737/98324163-82c92b00-2026-11eb-8b1f-d459e470e37a.png)

I've also noticed that the plugin does not read the d value for Revolute number 2 and number 3, you can change it to some absurd vlaue like 5000 and there'd be no difference.

Will need Vicente or Vincent to confirm this tho, just pointing out my observations.

**Comment by eduardsik** [comment]
@v-xup6 Thanks for you observation. I saw this shift in design on other staubli robots and it is involved in xml. I think somewhere else is the problem, but I can't find it.

**Comment by visose** [comment]
Hi, I recommend you find an image with dimensions like above for one of the existing Staubli robots in the Bartlett library, see what dimensions the numbers correspond to, and then replace them with the TX60 numbers of the same dimension.

**Comment by eduardsik** [comment]
I did this as you suggested and also calculated the DH parameters for the dimensions. As a result, the robot is much better assembled, but it always breaks down.  That's why I'm looking for your help.
I attached everything in the first comment.

**Comment by visose** [comment]
Strange, you shouldn't be calculating any dimensions, just copying the exact number from the drawing.

**Comment by eduardsik** [comment]
i checked once again and seems the data is good in xml. Could you please take a look at it? Maybe can ask you to create a sample sketch in grasshopper with TX60 to test it out and send to me as a reference? It would help a lot.

**Comment by visose** [comment]
Ok, I'll try to have a look in the evening. Just giving you a chance first, since it always feels better to figure out the issue on your own.

**Comment by eduardsik** [comment]
Thanks, I'll check one more time.

**Comment by eduardsik** [comment]
Did you manage to take a look at it?

**Comment by eduardsik** [comment]
Sorry but no success on my side. Could you manage to look at the files please?

**Comment by visose** [comment]
@eduardsik  I had a look at this and @v-xup6  is correct, that offset is not supported in the kinematics used by Robots (the d value of the second axis is not used). I don't think I will have time to work on adding this anytime soon, but would be a great if anyone else is willing to try.

**Comment by visose** [comment, marked as answer]
I will close this issue, the conclusion being, not a supported feature yet.

**Comment by visose** [comment]
@eduardsik, not sure if you still need this, almost two years later, but the upcoming release comes with a different inverse kinematics solver for spherical wrist robots that includes this offset. The TX60 seems to work fine:

<img width="413" alt="image" src="https://user-images.githubusercontent.com/5129614/184511448-1c140f8a-7cad-4f1b-ba7d-a4d12f51601e.png">

It was added as part of the support for Doosan robots, which also have this offset (this is sponsored work).

**Reply by eduardsik** [reply]
Great news! Thanks for the update.

**Comment by eduardsik** [comment]
Hello Vincent,

than you again to pointing me out this new version. I tried to control my robot but I have the same misalignment on the 4th axis. What am I doing wrong?
Here is the xml and the screenshot of the problem. 

May I ask you to help?

Thanks,
Eduard
```
<RobotCell name="Staubli TX60" manufacturer="Staubli">
    <Mechanisms group="0">
      <RobotArm model="TX60" manufacturer="Staubli" payload="4">
        <Base x="0.000" y="0.000" z="0.000" q1="1.000" q2="0.000" q3="0.000" q4="0.000"/>
        <Joints>
          <Revolute number="1" a ="0" d ="375" minrange = "-180" maxrange ="180" maxspeed ="435"/>
          <Revolute number="2" a ="290" d ="0" minrange = "-127.5" maxrange ="127.5" maxspeed ="410"/>
          <Revolute number="3" a ="0" d ="-20" minrange = "-142.5" maxrange ="142.5" maxspeed ="540"/>
          <Revolute number="4" a ="0" d ="310" minrange = "-270" maxrange ="270" maxspeed ="995"/>
          <Revolute number="5" a ="0" d ="0" minrange = "-122.5" maxrange ="133.5" maxspeed ="1065"/>
          <Revolute number="6" a ="0" d ="70" minrange = "-270" maxrange ="270" maxspeed ="1445"/>
        </Joints>
      </RobotArm>
    </Mechanisms>
    <IO>
      <DO names="BasicIO-1\%Q0,BasicIO-1\%Q1,BasicIO-1\%Q2"/>
      <DI names="BasicIO-1\%I0,BasicIO-1\%I1,BasicIO-1\%I2"/>
    </IO>
  </RobotCell>
```


https://www.dropbox.com/s/yxd5137oo9cdaof/Screenshot%20%282%29.png?dl=1

**Reply by eduardsik** [reply]
Could you please send me your model and xml?

**Reply by visose** [reply]
You almost have it, just flip the d values between joints number 2 and 3.

```
<Revolute number="2" a="290" d="-20" minrange="-127.5" maxrange="127.5" maxspeed="410" />
<Revolute number="3" a="0" d="0" minrange="-142.5" maxrange="142.5" maxspeed="540" />
```

**Reply by eduardsik** [reply]
Thanks, now almost good. Some disortion in the 6th joint appeared and the cplane is not in the center. How to fix these? 
[https://www.dropbox.com/s/pjjy8819yzxj7h0/Screenshot3.png?dl=1](url)

**Comment by visose** [comment]
Hi @eduardsik, see if this one works: [TX60.zip](https://github.com/visose/Robots/files/10198419/TX60.zip)

**Comment by eduardsik** [comment]
Thanks Vincent,

This is good now.

*Eduard Sik*
Creative Engineer <http://about.me/eduardsik> || eduardsik.com ||
***@***.*** || +36 70 237 4151 || book me
<https://doodle.com/eduardsik>
